### PR TITLE
Bug 1851676 - Adds ability to initialize Sentry on startup

### PR DESCRIPTION
**Pull request by boek**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1851676

### CODE REVIEW

File: android-components/components/lib/crash-sentry/src/main/java/mozilla/components/lib/crash/sentry/SentryService.kt [hunk 1]
```diff
@@ -34,6 +34,7 @@ import mozilla.components.concept.base.crash.Breadcrumb as MozillaBreadcrumb
  * @param sendEventForNativeCrashes Allows configuring if native crashes should be submitted. Disabled by default.
  * @param sentryProjectUrl Base URL of the Sentry web interface pointing to the app/project.
  * @param sendCaughtExceptions Allows configuring if caught exceptions should be submitted. Enabled by default.
+ * @param autoInitializeSentry Initializes Sentry
```
Comment (jonalmeida) [inline android-components/components/lib/crash-sentry/src/main/java/mozilla/components/lib/crash/sentry/SentryService.kt:37]:
```suggestion
 * @param autoInitializeSentry Initializes the Sentry SDK immediately on service creation.
```

File: docs/changelog.md [hunk 2]
```diff
@@ -10,10 +10,13 @@ permalink: /changelog/
...
 * [Configuration](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/.config.yml)
 
+* **lib-crash-sentry**
+  * Adds paramter to initialize Sentry with `SentryService`.
+
 * **ui-tabcounter**
   * Adds a mask overlay to the tabcounter that can be shown with `toggleCounterMask`.
 
-**feature-push**
+* **feature-push**
```
Comment (jonalmeida) [inline docs/changelog.md:19]: ty! 😄

**Review by jonalmeida — APPROVED** [review]
Offline, we discussed that we should still test out the performance impact of this change and let it bake in Nightly for a week before we consider uplifting.

File: docs/changelog.md [hunk 3]
```diff
@@ -10,10 +10,13 @@ permalink: /changelog/
 * [Gecko](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/plugins/dependencies/src/main/java/Gecko.kt)
 * [Configuration](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/.config.yml)
 
+* **lib-crash-sentry**
+  * Adds paramter to initialize Sentry with `SentryService`.
```
Comment (willdurand) [inline docs/changelog.md:14]: Typo:

```suggestion
  * Adds parameter to initialize Sentry with `SentryService`.
```

(sorry I only stumbled upon this patch accidentally, my OCD did the rest..)

**Comment by jonalmeida** [comment]
We discussed this patch some more offline: since we are going to initialize the Sentry SDK immediately, this is will then happen in [all processes](https://searchfox.org/mozilla-mobile/rev/caa06218324916873ec02c77a80ddf0821df167d/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#229-230): main and (Gecko) content.

There is no value in starting up the Sentry SDK in the content processes, because we do not send native crashes there any way, so we should try to initialize the Sentry SDK in only the main process if needed.

File: docs/changelog.md (hunk 3 above)
Comment (boek) [reply docs/changelog.md:14]: thank you!

**Comment by boek** [comment]
# Current head of main
release

# Adding Sentry to Startup
new

**Comment by boek** [comment]
# Moving initialization to `visualCompletenessQueue` ([a21378e](https://github.com/mozilla-mobile/firefox-android/pull/3502/commits/a21378e1cf5a25f34da86a8e13040bdb8cd4275b))
```
{'max': 797.0,
 'mean': 747.3,
 'median': 740.0,
 'min': 721.0,
 'replicate_count': 10,
 'replicates': [797.0, 778.0, 735.0, 732.0, 730.0, 744.0, 736.0, 744.0, 721.0,
                756.0],
 'stdev': 23.60343666126232}
 ```

**Comment by boek** [comment]
I am very confused
# Main ([9afb42e](https://github.com/mozilla-mobile/firefox-android/commit/9afb42efa4154f12ba40fb5428f07159bea12fb9))
```
{'max': 701.0,
 'mean': 645.53,
 'median': 643.0,
 'min': 614.0,
 'replicate_count': 100,
 'replicates': [686.0, 696.0, 629.0, 623.0, 629.0, 649.0, 664.0, 626.0, 619.0,
                647.0, 647.0, 633.0, 634.0, 642.0, 624.0, 653.0, 643.0, 674.0,
                617.0, 649.0, 639.0, 664.0, 640.0, 622.0, 633.0, 651.0, 648.0,
                622.0, 691.0, 629.0, 644.0, 623.0, 649.0, 628.0, 627.0, 634.0,
                643.0, 639.0, 631.0, 701.0, 633.0, 636.0, 650.0, 649.0, 642.0,
                626.0, 632.0, 646.0, 651.0, 638.0, 687.0, 644.0, 622.0, 656.0,
                660.0, 643.0, 614.0, 624.0, 634.0, 634.0, 647.0, 683.0, 653.0,
                636.0, 646.0, 641.0, 660.0, 636.0, 640.0, 631.0, 632.0, 642.0,
                675.0, 644.0, 653.0, 646.0, 647.0, 657.0, 641.0, 672.0, 638.0,
                652.0, 672.0, 662.0, 657.0, 628.0, 649.0, 655.0, 672.0, 642.0,
                639.0, 636.0, 649.0, 648.0, 695.0, 628.0, 629.0, 645.0, 660.0,
                652.0],
 'stdev': 18.24660200827773}
 ```
 
 # Initializing Sentry after visual completeness
 ```
{'max': 986.0,
 'mean': 923.08,
 'median': 918.0,
 'min': 896.0,
 'replicate_count': 100,
 'replicates': [958.0, 949.0, 906.0, 906.0, 933.0, 907.0, 897.0, 896.0, 906.0,
                909.0, 916.0, 930.0, 968.0, 932.0, 924.0, 909.0, 904.0, 932.0,
                910.0, 970.0, 928.0, 917.0, 935.0, 941.0, 908.0, 907.0, 911.0,
                911.0, 918.0, 908.0, 903.0, 931.0, 918.0, 910.0, 959.0, 924.0,
                912.0, 917.0, 916.0, 917.0, 901.0, 936.0, 916.0, 916.0, 921.0,
                965.0, 910.0, 920.0, 916.0, 929.0, 948.0, 912.0, 910.0, 917.0,
                913.0, 912.0, 967.0, 918.0, 907.0, 932.0, 908.0, 904.0, 913.0,
                941.0, 918.0, 937.0, 935.0, 960.0, 923.0, 908.0, 927.0, 911.0,
                901.0, 942.0, 918.0, 915.0, 914.0, 916.0, 955.0, 934.0, 920.0,
                917.0, 922.0, 930.0, 900.0, 915.0, 919.0, 940.0, 923.0, 986.0,
                928.0, 920.0, 919.0, 927.0, 927.0, 921.0, 908.0, 939.0, 919.0,
                929.0],
 'stdev': 17.655902953145986}
 ```

**Comment by boek** [comment]
# Main with `SENTRY_TOKEN` (df18d8c)
```
{'max': 1073.0,
 'mean': 1006.62,
 'median': 1004.5,
 'min': 972.0,
 'replicate_count': 100,
 'replicates': [1042.0, 1034.0, 999.0, 1010.0, 996.0, 979.0, 1021.0, 1005.0,
                1006.0, 982.0, 1002.0, 990.0, 1015.0, 1001.0, 1010.0, 1053.0,
                1014.0, 993.0, 989.0, 1006.0, 991.0, 1012.0, 1008.0, 1025.0,
                1001.0, 1006.0, 988.0, 1006.0, 988.0, 989.0, 1012.0, 979.0,
                1012.0, 994.0, 1059.0, 998.0, 985.0, 1006.0, 1011.0, 1035.0,
                986.0, 972.0, 992.0, 996.0, 998.0, 1029.0, 994.0, 999.0, 1007.0,
                993.0, 1018.0, 987.0, 992.0, 1014.0, 996.0, 984.0, 1039.0,
                1006.0, 1004.0, 1007.0, 1009.0, 991.0, 996.0, 999.0, 1000.0,
                999.0, 1014.0, 1029.0, 1011.0, 1026.0, 1016.0, 991.0, 1016.0,
                992.0, 1048.0, 999.0, 1009.0, 1021.0, 1073.0, 1003.0, 1010.0,
                1024.0, 1010.0, 993.0, 994.0, 996.0, 1001.0, 1000.0, 996.0,
                1036.0, 997.0, 1002.0, 1019.0, 997.0, 1005.0, 1012.0, 1010.0,
                1037.0, 989.0, 1027.0],
 'stdev': 17.809190681422688}
 ```
 
 # Initializing sentry after visual completeness (599785e)
```
{'max': 1091.0,
 'mean': 1008.01,
 'median': 1004.5,
 'min': 959.0,
 'replicate_count': 100,
 'replicates': [1025.0, 1050.0, 994.0, 975.0, 995.0, 981.0, 992.0, 959.0,
                1004.0, 997.0, 1011.0, 986.0, 1047.0, 1006.0, 990.0, 1000.0,
                1010.0, 1052.0, 997.0, 1004.0, 1024.0, 1002.0, 1021.0, 985.0,
                999.0, 1012.0, 988.0, 1012.0, 1051.0, 981.0, 997.0, 998.0,
                985.0, 997.0, 990.0, 1004.0, 1019.0, 1027.0, 1019.0, 987.0,
                980.0, 1019.0, 999.0, 993.0, 1025.0, 991.0, 1016.0, 999.0,
                1001.0, 1060.0, 1010.0, 1005.0, 1009.0, 1007.0, 1038.0, 989.0,
                1011.0, 989.0, 1016.0, 990.0, 1091.0, 1015.0, 1017.0, 995.0,
                992.0, 1036.0, 1011.0, 1002.0, 1017.0, 1024.0, 1001.0, 971.0,
                1014.0, 1009.0, 1014.0, 977.0, 1037.0, 982.0, 1004.0, 1025.0,
                998.0, 998.0, 1012.0, 997.0, 999.0, 993.0, 1015.0, 1071.0,
                1033.0, 977.0, 1012.0, 1013.0, 1044.0, 991.0, 1025.0, 1005.0,
                1001.0, 1015.0, 1026.0, 1022.0],
 'stdev': 21.612517416786716}
 ```

**Review by jonalmeida — APPROVED** [review]
This patch looks great, thank you for doing all the perf testing! Let's land this on Monday morning and let it bake for that week to ensure stability, before we request uplift where we will see the actual effects of this patch show up.

I ran this three times as well locally to validate the results too using e5807ea856 and baseline df18d8c513 with the perf tools command: `./measure_start_up.py -c10 nightly cold_main_first_frame <filename>`

<details>

<summary>Iteration 1</summary>

```
==> patch-a51-e5807ea856-0.txt <==
{'max': 1063.0,
 'mean': 969.8,
 'median': 962.0,
 'min': 917.0,
 'replicate_count': 10,
 'replicates': [1063.0, 985.0, 966.0, 917.0, 978.0, 962.0, 956.0, 962.0, 951.0,
                958.0]}

==> patch-a51-e5807ea856-pre-0.txt <==
{'max': 1080.0,
 'mean': 981.2,
 'median': 966.5,
 'min': 927.0,
 'replicate_count': 10,
 'replicates': [1080.0, 1035.0, 963.0, 957.0, 964.0, 927.0, 988.0, 947.0, 969.0,
                982.0]}

```

</details>


<details>

<summary>Iteration 2</summary>

```
==> patch-a51-e5807ea856-1.txt <==
{'max': 1038.0,
 'mean': 971.7,
 'median': 959.0,
 'min': 940.0,
 'replicate_count': 10,
 'replicates': [1038.0, 950.0, 940.0, 952.0, 996.0, 947.0, 967.0, 947.0, 1014.0,
                966.0]}

==> patch-a51-e5807ea856-pre-1.txt <==
{'max': 1015.0,
 'mean': 963.0,
 'median': 958.0,
 'min': 949.0,
 'replicate_count': 10,
 'replicates': [958.0, 950.0, 956.0, 949.0, 1015.0, 964.0, 958.0, 960.0, 969.0,
                951.0]}

```

</details>


<details>

<summary>Iteration 3</summary>

```
==> patch-a51-e5807ea856-2.txt <==
{'max': 1063.0,
 'mean': 972.6,
 'median': 958.5,
 'min': 934.0,
 'replicate_count': 10,
 'replicates': [936.0, 934.0, 952.0, 1063.0, 961.0, 938.0, 966.0, 1058.0, 962.0,
                956.0]}

==> patch-a51-e5807ea856-pre-2.txt <==
{'max': 1001.0,
 'mean': 956.6,
 'median': 952.0,
 'min': 929.0,
 'replicate_count': 10,
 'replicates': [986.0, 929.0, 961.0, 950.0, 954.0, 938.0, 936.0, 963.0, 948.0,
                1001.0]}

```

</details>

**Comment by jonalmeida** [comment]
> @[github-actions](https://github.com/apps/github-actions) github-actions bot removed the 🛬 needs landing (squash)

Ah, there's a ktlint failure which is why actions removed the label.

**Comment by boek** [comment]
https://github.com/Mergifyio backport releases_v118

**Pull request merged**